File: library/banner/src/main/java/com/example/banner/adapter/BannerPageAdapter.java
package com.example.banner.adapter;

import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.viewpager.widget.PagerAdapter;

import com.example.banner.holder.Holder;
import com.example.banner.view.BannerViewPager;

import java.util.List;

public class BannerPageAdapter<T> extends PagerAdapter {

    private BannerViewPager<T> viewPager;

    private Holder<T> holderCreator;

    protected List<T> data;

    private boolean canLoop = true;  // 是否支持无限循环

    private int childCount = 0;

    public void setCanLoop(boolean canLoop) {
        this.canLoop = canLoop;
    }

    @Override
    public int getCount() {
        int realCount = getRealCount();
        if (0 == realCount) return 0;
        if (canLoop) return Integer.MAX_VALUE;
        return realCount;
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
        return view == object;
    }

    public int getRealCount() {
        return data == null ? 0 : data.size();
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        int realPosition = getRealPosition(position);
        View view = holderCreator.createView(container.getContext());
        container.addView(view);
        //
        if (data != null && !data.isEmpty() && realPosition < data.size()) {
            holderCreator.updateUI(container.getContext(), view, realPosition, data.get(realPosition));
        }
        return view;
    }

    public int getRealPosition(int position) {
        int realCount = getRealCount();
        if (0 == realCount) return 0;
        if (canLoop) return position % realCount;
        return position;
    }

    public BannerPageAdapter(Holder<T> holderCreator, List<T> data) {
        this.holderCreator = holderCreator;
        this.data = data;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, @NonNull Object object) {
        View view = (View) object;
        container.removeView(view);
    }

    @Override
    public void finishUpdate(@NonNull ViewGroup container) {
        int position = viewPager.getCurrentItem();
        if (position == 0) {
            position = viewPager.getFirstItem();
        } else if (position == getCount() - 1) {
            position = viewPager.getLastItem();
        }
        try {
            viewPager.setCurrentItem(position, false);
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    public void setViewPager(BannerViewPager<T> viewPager) {
        this.viewPager = viewPager;
    }

    @Override
    public int getItemPosition(@NonNull Object object) {
        if (childCount > 0) {
            childCount--;
            return POSITION_NONE;
        }
        return super.getItemPosition(object);
    }

    @Override
    public void notifyDataSetChanged() {
        childCount = getCount();
        super.notifyDataSetChanged();
    }

    public void setData(List<T> data) {
        this.data = data;
        notifyDataSetChanged();
    }

}

File: library/banner/src/main/java/com/example/banner/adapter/BannerPageAdapter.kt
package com.example.banner.adapter

import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import com.example.banner.holder.Holder
import com.example.banner.view.BannerViewPager

class BannerPageAdapter<T>(private val holderCreator: Holder<T>, private var mData: List<T>?) : PagerAdapter() {

    private var viewPager: BannerViewPager<T>? = null

    private var canLoop = true // 是否支持无限循环

    private var childCount = 0

    fun setCanLoop(canLoop: Boolean) {
        this.canLoop = canLoop
    }

    override fun getCount(): Int {
        val realCount = getRealCount()
        if (0 == realCount) return 0
        return if (canLoop) Int.MAX_VALUE else realCount
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

    fun getRealCount(): Int {
        return mData?.size ?: 0
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val realPosition = getRealPosition(position)
        val view = holderCreator.createView(container.context)
        container.addView(view)
        //
        mData?.let { mData ->
            if (mData.isNotEmpty() && realPosition < mData.size) {
                holderCreator.updateUI(container.context, view, realPosition, mData[realPosition])
            }
        }
        return view
    }

    fun getRealPosition(position: Int): Int {
        val realCount = getRealCount()
        if (0 == realCount) return 0
        return if (canLoop) position % realCount else position
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        val view = `object` as View
        container.removeView(view)
    }

    override fun finishUpdate(container: ViewGroup) {
        viewPager?.let { viewPager ->
            var position = viewPager.currentItem
            if (position == 0) {
                position = viewPager.getFirstItem()
            } else if (position == count - 1) {
                position = viewPager.getLastItem()
            }
            try {
                viewPager.setCurrentItem(position, false)
            } catch (e: IllegalStateException) {
                e.printStackTrace()
            }
        }
    }

    fun setViewPager(viewPager: BannerViewPager<T>?) {
        this.viewPager = viewPager
    }

    override fun getItemPosition(`object`: Any): Int {
        if (childCount > 0) {
            childCount--
            return POSITION_NONE
        }
        return super.getItemPosition(`object`)
    }

    override fun notifyDataSetChanged() {
        childCount = count
        super.notifyDataSetChanged()
    }

    fun setData(data: List<T>?) {
        this.mData = data
        notifyDataSetChanged()
    }
}